### PR TITLE
feat (provider-utils): Add resolvable type and utility routine.

### PR DESCRIPTION
**Pull request by shaper**
Breakout from #3908 to land first.

### CODE REVIEW

File: packages/provider-utils/src/resolve.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+export async function resolve<T>(value: Resolvable<T>): Promise<T> {
+  // If it's a function, call it to get the value/promise
+  if (typeof value === 'function') {
+    const result = (value as Function)();
+    return Promise.resolve(result);
+  }
+
+  // Otherwise just resolve whatever we got (value or promise)
+  return Promise.resolve(value);
+}
```
Comment (lgrammel) [inline packages/provider-utils/src/resolve.ts:20]: to safe a few download bytes, I would do the following (more common pattern that I've seen):

```
if (typeof value === 'function') { 
  value = (value as Function)();
}

return Promise.resolve(value);
```

File: packages/provider-utils/src/resolve.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+// Helper for common case of resolving headers
+export type ResolvableHeaders = Resolvable<Record<string, string | undefined>>;
+export const resolveHeaders = (headers: ResolvableHeaders) => resolve(headers);
```
Comment (lgrammel) [inline packages/provider-utils/src/resolve.ts:24]: I don't know if `ResolvableHeaders` is needed. Do we have a `Headers` type somewhere (warning: might conflict with JS built-ins)

`resolveHeaders` is unnecessary, type inference will automatically do the right thing with just resolve.

Comment (lgrammel) [reply packages/provider-utils/src/resolve.ts:24]: (if `ResolvableHeaders` is needed, consider moving it to a separate file)

**Review by lgrammel — APPROVED** [review]

File: packages/provider-utils/src/resolve.ts (hunk 2 above)
Comment (lgrammel) [reply packages/provider-utils/src/resolve.ts:24]: Thought about it more: `ResolvableHeaders` has little value over `Resolvable<Headers>`. We should not include it imo (and instead think about a `Headers` type).

File: packages/provider-utils/src/resolve.ts (hunk 1 above)
Comment (shaper) [reply packages/provider-utils/src/resolve.ts:20]: Done, I had to add `value as T` to the `value` in the `return`.

**Pull request merged**